### PR TITLE
[EventGrid] Microsoft.AVS/privateClouds events

### LINKED ISSUE #26834: [Azure Event Grid - Event Grid Namespace ] API Review
New API Review meeting has been requested.

**Service Name**: Azure Event Grid - Event Grid Namespace 
**Review Created By**: Eugene Tolmachev
**Review Date**: 12/13/2023 09:00 AM PT
**Onboarding Record**: https://dev.azure.com/azure-sdk/Release/_workitems/edit/18565
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/24015
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: Add events schema for new 1st party/System topic - Microsoft.AVS/privateClouds

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=5f31d015-4c8a-ee11-8179-000d3a341a24)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by et1975**
# Data Plane API - Pull Request
Onboard Microsoft.AVS/privateClouds as system topic Event Grid source.

## API Info: The Basics

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
* Design Document: https://microsoft.sharepoint.com/:w:/t/AzureDedicatedSkylink/Ebb0SQIcowtOgK880FfYP2wB-i3j7WizYwvIgW20JmIyzQ?e=v3H9zD
* Previous API Spec Doc:
* Updated paths:

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/26834

### CODE REVIEW

File: specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01/examples/cloud-events-schema/script_execution_failed.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "source": "/subscriptions/{subscription_id}/resourceGroups/{resource_group_name}/providers/Microsoft.AVS/privateClouds/{sddc_name}",
+  "subject": "/scriptExecutions/{execution_name}",
+  "type": "Microsoft.AVS.ScriptExecutionFailed",
+  "time": "2023-12-09T13:50:19.9995668-08:00",
+  "id": "9b87886d-21a5-4af5-8e3e-10c4b8dac73b",
+  "data": {
+    "operationId": "fe1cb595-71c8-47d7-b505-60d39a44d53d",
+    "failure": "05/04/2023 03:15:40 Get-SpbmStoragePolicy SpbmStoragePolicy with name ' RAID-5 FTT-1' was not found using the specified filter(s).",
```
Comment (boumenot) [inline specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01/examples/cloud-events-schema/script_execution_failed.json:9]: Is it really "' RAID-5 FTT-1" or is that an extra leading space from you?

File: specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01/examples/event-grid-schema/cluster_updated.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    ],
+    "removed": [
+      "esx15-r01.p01.431cc…"
+    ],
+    "maintenance": [
+      "esx01-r03.p01.431cc…"
+    ],
+    "operationId": "c8c0b304-2db3-402b-8a6b-06065e0e1755"
+  },
+  "subject": "/clusters/{cluster_name}",
```
Comment (boumenot) [inline specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01/examples/event-grid-schema/cluster_updated.json:17]: This may have been in the sample code, but is the caller expected to parse subject to get cluster_name, subscription_id, resource_group_name, and sddc_name?  Do you consider this to be regular enough to make it straightforward, or should they be well-defined nouns in the system.

Or, is this the existing style of other Event Grids, and we're following an established pattern.

**Review by boumenot — APPROVED** [review]

File: specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01/examples/cloud-events-schema/script_execution_failed.json (hunk 1 above)
Comment (et1975) [reply specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01/examples/cloud-events-schema/script_execution_failed.json:9]: This is an (almost verbatim) example of the error we've reported 2 weeks ago - we don't trim spaces apparently :)

**Comment by et1975** [comment]
@Kishp01 I don't see existing specs in eventgrid linking to the examples, it's also not even possible to map the schema to an example because the schema only defines `data` element, not the entire event captured in the example.

I also don't see how I can exclude the examples, please advise.

File: specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+      "allOf": [
+        {
+          "$ref": "#/definitions/AVSScriptExecutionEventProperties"
+        }
+      ],
+      "properties": {
+        "namedOutputs": {
+          "description": "Named outputs of completed execution, if any.",
+          "type": "object",
+          "items": {
```
Comment (et1975) [inline specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json:122]: Josh, you mentioned there's some decoration we can apply to make the property generate as `Dictionary<string,string>`?

File: specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    },
+    "AVSClusterEventProperties": {
+      "description": "Schema of the Data property of an EventGridEvent for Microsoft.AVS/clusters events.",
+      "type": "object",
+      "properties": {
+        "operationId": {
+          "description": "Id of the operation that caused this event.",
+          "type": "string"
+        },
+        "hostsAdded": {
```
Comment (et1975) [inline specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json:171]: we've had some back and forth about the naming here, I'd rather not use "name", as in "hostNamesAdded" because we are not adding names, we're adding hosts. Normally I'd model this with the type system, defining `FQDN` to use instead of string, regardless... imo anyone using this event will not be under any illusion that this is a collection of hosts.

File: specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        }
+      }
+    },
+    "AVSClusterEventProperties": {
+      "description": "Schema of the Data property of an EventGridEvent for Microsoft.AVS/clusters events.",
+      "type": "object",
+      "properties": {
+        "operationId": {
+          "description": "Id of the operation that caused this event.",
+          "type": "string"
```
Comment (et1975) [inline specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json:169]: There's no API that treats this UUID, so I'd like to keep it as string.

File: specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+      "description": "Schema of the Data property of an EventGridEvent for a Microsoft.AVS.PrivateCloudUpdating event.",
+      "type": "object",
+      "allOf": [
+        {
+          "$ref": "#/definitions/AVSPrivateCloudEventProperties"
+        }
+      ],
+      "properties": {}
+    },
+    "AVSPrivateCloudUpdatedEventData": {
```
Comment (JoshLove-msft) [inline specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json:20]: Why does this have both AVS and PrivateCloud? I thought we were just going to do AVS. Also, it should be cased as "Avs"

File: specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json (hunk 4 above)
Comment (JoshLove-msft) [reply specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json:171]: I think "addedHostNames" addresses the confusion around adding hosts vs names.

File: specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json (hunk 6 above)
Comment (et1975) [reply specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json:20]: AVS is the offering, PrivateCloud is the resource type and the subject of the event. 
I believe the agreement was to prefix every type with AVS to distinguish the from other potential cluster, privateCloud, scriptExecution, etc.

Comment (JoshLove-msft) [reply specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json:20]: Sounds good

File: specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json (hunk 4 above)
Comment (et1975) [reply specification/eventgrid/data-plane/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json:171]: I don't believe it does address the confusion, as "hostNames" is not what is affected by the event, but the cluster's hosts composition. The "name" itself is not sufficiently precise either and potentially bigger source of ambiguity here, but I'm sure we all have bigger fish to fry than argue about this.

File: specification/eventgrid/data-plane/readme.md [hunk 7]
```diff
@@ -333,13 +333,14 @@ input-file:
...
 ```
 
 If there are files that should not be in the `all-api-versions` set, 
 uncomment the  `exclude-file` section below and add the file paths.
 
 ``` yaml $(tag) == 'all-api-versions'
-#exclude-file: 
-#  - $(this-folder)/Microsoft.Example/stable/2010-01-01/somefile.json
+exclude-file: 
+ - $(this-folder)/Microsoft.EventGrid/preview/2023-06-01-preview/examples
```
Comment (JoshLove-msft) [inline specification/eventgrid/data-plane/readme.md:344]: Why was this changed?

File: specification/eventgrid/data-plane/readme.md [hunk 8]
```diff
@@ -333,13 +333,14 @@ input-file:
   - $(this-folder)/Microsoft.ResourceNotifications/stable/2018-01-01/common.json
   - $(this-folder)/Microsoft.ResourceNotifications/stable/2018-01-01/HealthResources.json
   - $(this-folder)/Microsoft.ResourceNotifications/stable/2018-01-01/Resources.json
+  - $(this-folder)/Microsoft.AVS/preview/2023-09-01-preview/PrivateCloud.json
```
Comment (JoshLove-msft) [inline specification/eventgrid/data-plane/readme.md:336]: Please use the existing stable/2018-01-01 pattern. Also I think this entry needs to be added in the sections above as well.

File: specification/eventgrid/data-plane/readme.md (hunk 7 above)
Comment (et1975) [reply specification/eventgrid/data-plane/readme.md:344]: I tried following the instructions to avoid triggering Avocado failure, I'll undo the change.

**Comment by et1975** [comment]
Addressed the comments.

**Review by JoshLove-msft — APPROVED** [review]
